### PR TITLE
Jon/fix/balance-loading

### PR DESCRIPTION
**Pull request by Jon-edge**
https://github.com/EdgeApp/edge-react-gui/assets/90650827/5cf0c3e2-ce75-4a24-a386-5a1210aa7516

### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

 none

### Requirements

If you have made **any** visual changes to the GUI. Make sure you have:

- [x] Tested on iOS device
- [x] Tested on Android device
- [ ] Tested on small-screen device (iPod Touch)
- [ ] Tested on large-screen device (tablet)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207604271105270

### CODE REVIEW

File: src/components/ui4/BalanceCardUi4.tsx [hunk 1]
```diff
@@ -64,26 +68,31 @@ export const BalanceCardUi4 = (props: Props) => {
...
   const animateNumber = fiatAmount < MAX_ANIMATED_AMOUNT
 
+  useAsyncEffect(
+    async () => {
+      await account.waitForAllWallets()
+      setIsWalletsLoaded(true)
+    },
+    [],
+    'BalanceCard'
+  )
```
Comment (swansontec) [inline src/components/ui4/BalanceCardUi4.tsx:86]: I don't think we need `useAsyncEffect` here. Instead, we can consolidate the `exchangeRatesReady` logic with the `isWalletsLoaded` logic into a single loop:

```ts
const activeWalletIds = useWatch(account, 'activeWalletIds')
const currencyWallets = useWatch(account, 'currencyWallets')
const currencyWalletErrors = useWatch(account, 'currencyWalletErrors')

const exchangeRatesReady = React.useMemo(() =>
  activeWalletIds.every(walletId => {
    // Ignore wallets that have crashed:
    if (currencyWalletErrors[walletId]) return true

    // Both the wallet and its rate need to be loaded:
    const wallet = currencyWallets[walletId]
    return wallet != null && exchangeRates[`${wallet.currencyInfo.currencyCode}_${defaultIsoFiat}`] != null
  })
, [activeWalletIds, currencyWallets, currencyWalletErrors])
```

File: src/hooks/useAccountWalletsSyncProgress.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          currencyWalletErrors[walletId] != null
+        )
+      }).length,
+    [currencyWalletErrors, progressMap, syncableWalletIds]
+  )
+
+  const progress = React.useMemo(
+    () => (syncableWalletIds.length === 0 ? 100 : 100 * (syncedWallets / syncableWalletIds.length)),
+    [syncableWalletIds.length, syncedWallets]
+  )
```
Comment (swansontec) [inline src/hooks/useAccountWalletsSyncProgress.tsx:67]: Don't use a memo here. The `useMemo` hook needs to loop over its dependency array to see if anything has changed, which is going to cost more than just re-doing the division on each run. Math is really fast.

Comment (swansontec) [reply src/hooks/useAccountWalletsSyncProgress.tsx:67]: Also, I request that this hook return a value in the range of 0-1, and have the name `useAccountSyncRatio`. That makes it clearer what the units are, and mirrors the fact that the wallets themselves operate on a 0-1 scale. Visual components like the sync bar can convert the range as necessary for display purposes (like 0% - 100%).

**Review by swansontec — CHANGES_REQUESTED** [review]

**Review by paullinator — APPROVED** [review]

**Pull request merged**